pkg/test/framework/components/mixer/mixer.go

```diff
@@ -0,0 +1,213 @@
+//  Copyright 2018 Istio Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package mixer
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"testing"
+	"time"
+
+	"go.uber.org/multierr"
+	"google.golang.org/grpc"
+
+	istio_mixer_v1 "istio.io/api/mixer/v1"
+	"istio.io/istio/mixer/adapter"
+	"istio.io/istio/mixer/pkg/attribute"
+	"istio.io/istio/mixer/pkg/server"
+	generatedTmplRepo "istio.io/istio/mixer/template"
+	"istio.io/istio/pkg/test/framework/environment"
+	"istio.io/istio/pkg/test/framework/environments/kubernetes"
+	"istio.io/istio/pkg/test/framework/environments/local"
+	"istio.io/istio/pkg/test/framework/internal"
+	"istio.io/istio/pkg/test/kube"
+)
+
+type deployedMixer struct {
+	// Indicates that the component is running in local mode.
+	local bool
+
+	conn   *grpc.ClientConn
+	client istio_mixer_v1.MixerClient
+
+	args    *server.Args
+	server  *server.Server
+	workdir string
+
+	forwarder *kube.PortForwarder
+}
+
+var _ environment.DeployedMixer = &deployedMixer{}
+var _ internal.Configurable = &deployedMixer{}
+var _ io.Closer = &deployedMixer{}
+
+// InitLocal initializes a new Mixer component for the local environment.
+func InitLocal(ctx environment.ComponentContext) (interface{}, error) {
```

Comment:
I'm not sure I like mixing all environments in the same file.  Could we have separate components for local/kube? I suppose a structure like `/pkg/test/framework/component/mixer/<local|kube>/` would work.

Reply:
This is a per-component decision, I think. For components such as Mixer, there is a huge overlap in terms of functionality. Keeping them colocated helps with maintenance. For other components (such as Pilot) it probably makes sense to have environment specific implementations. The current model allows both approaches.